src/freshrssapi.cpp

```diff
@@ -0,0 +1,546 @@
+{
+	std::string* pbuf = static_cast<std::string*>(userp);
+	pbuf->append(static_cast<const char*>(buffer), size * nmemb);
+	return size * nmemb;
+}
+
+std::string FreshRssApi::retrieve_auth()
+{
+	CURL* handle = curl_easy_init();
+	Credentials cred = get_credentials("freshrss", "FreshRSS");
+	if (cred.user.empty() || cred.pass.empty()) {
+		return "";
+	}
+
+	char* username = curl_easy_escape(handle, cred.user.c_str(), 0);
+	char* password = curl_easy_escape(handle, cred.pass.c_str(), 0);
+
+	std::string postcontent = strprintf::fmt(
+			"service=reader&Email=%s&Passwd=%s&source=%s%2F%s&accountType="
+			"HOSTED_OR_GOOGLE&continue=http://www.google.com/",
+			username,
+			password,
+			PROGRAM_NAME,
+			utils::program_version());
+
+	curl_free(username);
+	curl_free(password);
+
+	std::string result;
+
+	utils::set_common_curl_options(handle, cfg);
+	curl_easy_setopt(handle, CURLOPT_WRITEFUNCTION, my_write_data);
+	curl_easy_setopt(handle, CURLOPT_WRITEDATA, &result);
+	curl_easy_setopt(handle, CURLOPT_POSTFIELDS, postcontent.c_str());
+	curl_easy_setopt(handle,
+		CURLOPT_URL,
+		(cfg->get_configvalue("freshrss-url") + FRESHRSS_LOGIN).c_str());
+	curl_easy_perform(handle);
+	curl_easy_cleanup(handle);
+
+	for (const auto& line : utils::tokenize(result)) {
+		LOG(Level::DEBUG, "FreshRssApi::retrieve_auth: line = %s", line);
+		if (line.substr(0, 5) == "Auth=") {
+			std::string auth = line.substr(5, line.length() - 5);
+			return auth;
+		}
+	}
+
+	return "";
+}
+
+std::vector<TaggedFeedUrl> FreshRssApi::get_subscribed_urls()
+{
+	std::vector<TaggedFeedUrl> urls;
+
+	CURL* handle = curl_easy_init();
+	std::string result;
+	curl_slist* custom_headers{};
+	add_custom_headers(&custom_headers);
+	curl_easy_setopt(handle, CURLOPT_HTTPHEADER, custom_headers);
+
+	utils::set_common_curl_options(handle, cfg);
+	curl_easy_setopt(handle, CURLOPT_WRITEFUNCTION, my_write_data);
+	curl_easy_setopt(handle, CURLOPT_WRITEDATA, &result);
+	curl_easy_setopt(handle,
+		CURLOPT_URL,
+		(cfg->get_configvalue("freshrss-url") + FRESHRSS_SUBSCRIPTION_LIST)
+		.c_str());
+	curl_easy_perform(handle);
+	curl_easy_cleanup(handle);
+	curl_slist_free_all(custom_headers);
+
+	LOG(Level::DEBUG,
+		"FreshRssApi::get_subscribed_urls: document = %s",
+		result);
+
+	json_object* reply = json_tokener_parse(result.c_str());
+	if (reply == nullptr) {
+		LOG(Level::ERROR,
+			"FreshRssApi::get_subscribed_urls: failed to parse "
+			"response "
+			"as JSON.");
+		return urls;
+	}
+
+	json_object* subscription_obj{};
+	json_object_object_get_ex(reply, "subscriptions", &subscription_obj);
+	array_list* subscriptions = json_object_get_array(subscription_obj);
+
+	int len = array_list_length(subscriptions);
+
+	for (int i = 0; i < len; i++) {
+		std::vector<std::string> tags;
+		json_object* sub =
+			json_object_array_get_idx(subscription_obj, i);
+
+		json_object* id_str{};
+		json_object_object_get_ex(sub, "id", &id_str);
+		const char* id = json_object_get_string(id_str);
```

Comment:
Suggestion: move this down, closer to `escaped_id`.

Reply:
This section was taken wholesale from the feedhq api. I'll change my version but I'll leave that alone for now.